aip/general/0236.md

```diff
@@ -0,0 +1,351 @@
+---
+id: 236
+state: approved
+created: 2023-03-30
+updated: 2023-03-30
+placement:
+  category: resource-design
+  order: 0
+---
+
+# Policy preview
+
+A policy is a resource that provides rules that admit or deny access to other
+resources. Generally, the outcome of a policy can be evaluated to a specific set
+of outcomes.
+
+Changes to policies without proper validation may have unintended consequences
+that can severely impact a customer’s overall infrastructure setup. To safely
+update resources, it is beneficial to test these changes via policy rollout 
+APIs.
+
+Preview is a rollout safety mechanism for policy resources, which gives the
+customer the ability to validate the effect of their proposed changes against
+production traffic prior to the changes going live. The result of the policy
+evaluation against traffic is logged in order to give the customer the data
+required to test the correctness of the change.
+
+Firewall policies exemplify a case that is suitable for previewing. A new 
+configuration can be evaluated against traffic to observe which IPs would be 
+allowed or denied. This gives the customer the data to guide a decision on
+whether to promote the proposed changes to live.
+
+The expected flow for previewing a policy is as follows:
+
+1. The user creates an experiment containing a new policy configuration
+   intended to replace the live policy.
+2. The user uses the "preview" method to start generating logs which compare
+   the live and experiment policy evaluations against live traffic.
+3. The user inspects the logs to determine whether the experiment has the
+   intended result.
+4. The user uses the "commit" method to promote the experiment to live.
+
+## Guidance
+
+### Non-goals
+
+This proposal is for a safety mechanism for policy rollouts only. Safe rollouts
+for non-policy resources are not in scope.
+
+### Experiments
+
+A new configuration of a policy to be previewed is stored in a nested
+collection under the policy called experiments.
+
+A hypothetical policy resource called, `Policy`, is used throughout. It has the
+following resource name pattern:
+
+`projects/{project}/locations/{location}/policies/{policy}`
+
+The experimental versions of the resource used for previewing or other safe
+rollout practices are represented as a nested collection under `Policy` using a
+new resource type. The resource type **must** follow the naming convention
+*RegularResourceType*`Experiment`. 
+
+The following pattern is used for the experiment collection:
+
+`projects/{project}/locations/{location}/policies/{policy}/experiments/{experiment}`
```

Comment:
Should this be,

`projects/{project}/locations/{location}/policies/{policy}/previews/{preview}`

because you might have

`projects/{project}/locations/{location}/policies/{policy}/rollouts/{rollouts}`

for the same policy resource?